kv_storage.h

```diff
@@ -0,0 +1,230 @@
+    listener_.mq_.EmplaceMessage(new MQMessageGet(key, std::move(pr)));
+    return future;
+  }
+
+  typename T_ENTRY::Value Get(typename T_ENTRY::Key&& key) {
+    return AsyncGet(std::forward<typename T_ENTRY::Key>(key)).get();
+  }
+
+  std::future<void> AsyncSet(const T_ENTRY& entry) {
+    std::promise<void> pr;
+    std::future<void> future = pr.get_future();
+    // The order of the next two calls is important for eventual consistency.
+    listener_.mq_.EmplaceMessage(new MQMessageSet(entry, std::move(pr)));
+    stream_.Publish(entry);
+    return future;
+  }
+
+  void Set(const T_ENTRY& entry) { AsyncSet(entry).wait(); }
+
+  // For testing purposes.
+  bool CaughtUp() const { return listener_.caught_up_; }
+  size_t EntriesSeen() const { return listener_.entries_seen_; }
+
+ private:
+  // Stateful storage.
+  struct Storage {
+    std::unordered_map<typename T_ENTRY::Key, typename T_ENTRY::Value, typename T_ENTRY::Key::HashFunction>
+        data;
+  };
+
+  // The logic to "interleave" updates from Sherlock stream with inbound KeyValueStorage requests.
+  struct MQMessage {
+    virtual void DoIt(Storage& storage) = 0;
+  };
+
+  struct MQMessageEntry : MQMessage {
+    // TODO(dkorolev): A single entry is fine to copy, polymorphic ones should be std::move()-d.
+    T_ENTRY entry;
+
+    explicit MQMessageEntry(const T_ENTRY& entry) : entry(entry) {}
+
+    virtual void DoIt(Storage& storage) { storage.data[entry.key] = entry.value; }
+  };
+
+  struct MQMessageGet : MQMessage {
+    typename T_ENTRY::Key key;
+    std::promise<typename T_ENTRY::Value> pr;
+
+    explicit MQMessageGet(const typename T_ENTRY::Key& key, std::promise<typename T_ENTRY::Value>&& pr)
+        : key(key), pr(std::move(pr)) {}
+
+    virtual void DoIt(Storage& storage) {
+      const auto cit = storage.data.find(key);
+      if (cit != storage.data.end()) {
+        pr.set_value(cit->second);
+      } else {
+        pr.set_exception(std::make_exception_ptr(KeyNotPresentException()));
+      }
+    }
+  };
+
+  struct MQMessageSet : MQMessage {
+    T_ENTRY e;
+    std::promise<void> pr;
+
+    explicit MQMessageSet(const T_ENTRY& e, std::promise<void>&& pr) : e(e), pr(std::move(pr)) {}
+
+    // Important note: The entry added will eventually reach the storage via the stream.
+    // Thus, in theory, `MQMessageSet::DoIt()` could be a no-op.
+    // This code still updates the storage, to have the API appear more lively to the user.
+    // Since the actual implementation of `Set` pushes the `MQMessageSet` message before publishing
+    // an update to the stream, the final state will always be [evenually] consistent.
+    // The practical implication here is that an API `Get()` after an api `Set()` may and will return data,
+    // that might not yet have reached the storage, and thus relying on the fact that an API `Get()` call
+    // reflects updated data is not reliable from the point of data synchronization.
```

Comment:
Then why not wait for the data to reach the storage in a natural way?

My mind is currently Flux/React-biased, thus I like the idea to have `Get`s always return the current state, be it transient or final. I.e. if the key is being `Set` but not yet settled, `Get` should return undefined or previous state until the actual change event is reported by the stream.
This could certainly be a complication if the KV Storage is reliable enough to have the value delivered to the storage in any conditions (I'm not sure it is).